ipc/provider/src/dryrun.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: MIT
+
+//! Performs dry run for the ipc provider instead of directly submitter the txn on chain
+
+use crate::preflight::Preflight;
+use anyhow::anyhow;
+use base64::Engine;
+use ethers_contract::core::abi::{Function, Tokenizable};
+use ethers_contract::encode_function_data;
+use fvm_ipld_encoding::{BytesSer, RawBytes};
+use ipc_actors_abis::register_subnet_facet;
+use ipc_api::subnet::ConstructParams;
+use std::collections::BTreeMap;
+
+pub struct IPCDryRunProvider {
+    pub(crate) preflight: Preflight,
+}
+
+impl IPCDryRunProvider {
+    pub fn create_subnet(&self, params: ConstructParams) -> anyhow::Result<()> {
+        let params = self.preflight.create_subnet(params)?;
+        let params = register_subnet_facet::ConstructorParams::try_from(params)?;
+
+        let data = to_fvm_calldata(
+            &register_subnet_facet::REGISTERSUBNETFACET_ABI.functions,
+            "newSubnetActor",
+            params,
+        )?;
```

Comment:
This suggests this dryrun functionality should live under the `evm` module.

Reply:
it's kind of awkward in the sense that the calldata for fvm is derived from evm.

Reply:
it has to go through evm first, then convert to fvm.

Reply:
The `params` for the `Message` are indeed derived from the `calldata` of the EVM transaction request. However the fact that this even knows that it's calling one of the IPC Solidity contracts means it knows that the subnet is implemented by `fevm` (the only type currently, but not necessarily forever), so _this_ particular thing would _only_ work as a generic parameter to print _iff_ we already know the subnet is implemented with Solidity, therefore _this_ dryrun implementation should also be under the `evm` namespace.